src/pybind/mgr/dashboard/frontend/src/app/shared/directives/update-scope.directive.ts

```diff
@@ -0,0 +1,34 @@
+import { AfterViewInit, Directive, ElementRef, Input, Optional } from '@angular/core';
+import { Permissions } from '../models/permissions';
+import { AuthStorageService } from '../services/auth-storage.service';
+import { PermissionScopeDirective } from './permission-scope.directive';
+
+@Directive({
+  selector: 'input:not([cdNoUpdateScope]), select:not([cdNoUpdateScope]), [cdUpdateScope]'
+})
+export class UpdateScopeDirective implements AfterViewInit {
+  private permissions: Permissions;
+  private service_name: keyof Permissions;
+
+  @Input() disabled: boolean;
+
+  constructor(
+    @Optional() private scope: PermissionScopeDirective,
+    private authStorageService: AuthStorageService,
+    private elementRef: ElementRef
+  ) {
+    this.permissions = this.authStorageService.getPermissions();
+  }
+
+  ngAfterViewInit() {
+    if (this.scope !== null) {
+      this.service_name = this.scope.cdScope;
+    }
+    if (
+      this.service_name !== undefined &&
+      (this.disabled || !this.permissions[this.service_name].update)
+    ) {
+      this.elementRef.nativeElement.disabled = true;
+    }
+  }
+}
```

Comment:
```suggestion
import { AfterViewInit, Directive, ElementRef, Input } from '@angular/core';

import { Permissions } from '../models/permissions';
import { AuthStorageService } from '../services/auth-storage.service';

@Directive({
  selector: 'form[data-scope] input, form[data-scope] select, [cdUpdateScope]'
})
export class UpdateScopeDirective implements AfterViewInit {
  private permissions: Permissions;
  private service_name: keyof Permissions;

  constructor(private authStorageService: AuthStorageService, private elementRef: ElementRef) {}

  ngAfterViewInit() {
    const parentForm = this.elementRef.nativeElement.closest('form');
    this.service_name = parentForm.dataset.scope;
    if (this.service_name) {
      this.permissions = this.authStorageService.getPermissions();

      if (!this.permissions[this.service_name].update) {
        this.elementRef.nativeElement.disabled = true;
      }
    }
  }
}

```

Reply:
> We can use a data attribute instead of a custom directive to define the scope of the form, then we can read that in a similar way as we did the directive value.

The reason we avoided relying on direct DOM manipulation with `elementRef.nativeElement` is that in Angular [this is regarded as a risky and unsafe practice and only documented as a "last resort" method](https://angular.io/api/core/ElementRef#properties). That's why here we're using Templates and (dummy) attribute binding instead.

> Although I would still prefer to have 1 directive that would be used in the form, I will accept the PR with the directive on the form control level, I will try 1 last time to remove the dummy directive.

I don't see why we can only use a single directive here. We somehow need to 'tag' the higher-level feature scope (e.g.: OSD, RBD, etc.), but still need another way to 'tag' which lower-level elements we want to be affected by the higher-level scope features. Angular DOM queries only work with Template vars and Angular Directives (yeah, and `ElementRef`, but that's what we are working around), and something similar happens to the opposite approach: Component/Directive Dependency Injection.

Reply:
> The reason we avoided relying on direct DOM manipulation with elementRef.nativeElement is that in Angular this is regarded as a risky and unsafe practice and only documented as a "last resort" method. That's why here we're using Templates and (dummy) attribute binding instead.

Got it. I will aprove the PR with 2 directives so we don't have to resort to this.

> We somehow need to 'tag' the higher-level feature scope (e.g.: OSD, RBD, etc.), but still need another way to 'tag' which lower-level elements we want to be affected by the higher-level scope features. 

Do we plan to use this directives in any other forms? 
In all the forms where we are using this directives we are always applying it to all inputs.
In the current code it's not even possible to specify which input should be disabled or not, or select a different scope.

Reply:
Currently, I wrote this directive for disabling the inputs which are at present useless when switched to readonly mode (Osd flags Component, Osd recovery speed component, osd pg scrub component). But with a few tweaks in the future, we can improve this code by specifying which input to be disabled or selecting multiple scopes. @tspmelo